### ISSUE TITLE
Metrics instrumentation dbapi

### ISSUE THREAD

**Issue by srikanthccv** [open-telemetry/opentelemetry-python-contrib#1158]
DB metrics: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/database-metrics.md

**Comment by rblejter** [comment]
Can you assign this issue to me?

**Comment by avzis** [comment]
Hi, can you assign this to me?
Thanks!

**Comment by srikanthccv** [comment]
Feel free to ask any questions here or on slack.

**Comment by tammy-baylis-swi** [comment]
I think this issue is still relevant. The db semconv for metrics is now at: https://github.com/open-telemetry/semantic-conventions/blob/main/model/db/metrics.yaml

I think DB-API is generating these metrics currently:
* `db.client.operation.duration`
* `db.client.response.returned_rows`

But does not yet generate these:

1. `db.client.connection.count`
2. `db.client.connection.idle.max`
3. `db.client.connection.idle.min`
4. `db.client.connection.max`
5. `db.client.connection.pending_requests`
6. `db.client.connection.timeouts`
7. `db.client.connection.create_time`
8. `db.client.connection.wait_time`
9. `db.client.connection.use_time`

DB-API instrumentation is a sort of core util used by other instrumentors such as SqlAlchemy and psycopg(2). I think all of the above are relevant for each "concrete" framework instrumentation but whoever works on this should double-check responsibility.

**Comment by ahmed5145** [comment]
Picking this up after the recent re-triage. Thanks @tammy-baylis-swi for the breakdown. Before writing any code I want to nail down who owns what, since that responsibility question is the whole thing here.

Where `dbapi` core stands today: it already emits the two operation-level metrics.

- `db.client.operation.duration`
- `db.client.response.returned_rows`

The 9 metrics still on the list are all connection-pool metrics: `connection.count`, `idle.max`, `idle.min`, `max`, `pending_requests`, `timeouts`, `create_time`, `wait_time`, `use_time`. Raw DB-API (PEP 249) has no pool. It wraps a single `connect()` and its cursors. There is nothing in the core util that can observe a checkout, an overflow, or a timeout, so I don't think these can come from `dbapi` at all.

So my read on the split:

- `dbapi` core stays at the operation level, which is already done. The pool metrics don't belong here.
- Pool and connection metrics come from the concrete pooled instrumentations (SQLAlchemy, psycopg, and so on), where an actual pool exists to watch.

One concrete starting point: SQLAlchemy already reports connection usage, but under the deprecated `db.client.connections.usage` name (`connections_usage`) rather than the current `db.client.connection.count`. Migrating that to `connection.count` with `state=idle`/`used`, then adding the other pool metrics SQLAlchemy can see from its pool events, is a well-scoped first task.

Plan I'd suggest:

1. Call `dbapi` core done for the metrics that actually apply to it.
2. Handle pool and connection metrics per concrete instrumentation, starting with the SQLAlchemy migration to `connection.count`.

Does that match what you had in mind? If so I'm happy to take the SQLAlchemy piece first. Also worth deciding whether the two earlier PRs (#1620, #1647, both closed) should stay closed. They predate the current metrics code, so I don't think there's anything to salvage from them.

**Comment by tammy-baylis-swi** [comment]
Hi @ahmed5145 , thanks for that!

I've created a new, separate issue https://github.com/open-telemetry/opentelemetry-python-contrib/issues/5060 with your great find about the deprecated SQLAlchemy instrumentor metrics, since it ties into semconv stabilization and it's a well-scoped find on your part. I can't seem to directly assign it to you right now, so I've cc'd you on it in comment if you're interested still in taking it up.

For the `dbapi` vs other scope, generally I agree about responsibility of concrete connection. The `dbapi` instrumentor is historically an interesting package that's more of a util to decrease duplication but can be difficult to update. Therefore, I say go for it at least in a PoC e.g. start with other pool metrics in sqlalchemy first. If later on there are duplications between the other instrumentors as they get their metrics, then a design could be worked on to consolidate.

**Comment by ahmed5145** [comment]
Hi @tammy-baylis-swi , thank you for splitting this and for the steer.

That lines up with how I was thinking about it: dbapi core stays at the operation level since raw DB-API has no pool, and the pool/connection metrics belong in the concrete pooled instrumentors where there's an actual pool to observe. SQLAlchemy is a good starting point for that. I agree with you, if duplication across the pooled instrumentors turns up later, we can revisit consolidating the shared pieces then.

I've picked up #5060 and will start on the `connection.count` migration there.
Keeping this issue as the umbrella for the remaining pool metrics.

**Comment by ahmed5145** [comment]
@tammy-baylis-swi #5063 is up for `db.client.connection.count` and awaiting review.  Thank you for the review in advance.

On the remaining pool metrics for SQLAlchemy, plus a few calls, I'd like your read on before I open or do anything.

Remaining slices (each independently reviewable):

- [ ] `connection.max` + `connection.idle.max`: static gauges read from pool config (`pool_size`, `max_overflow`). One PR.
- [ ] `connection.use_time`: histogram off the existing checkout/checkin listeners. Stash the checkout timestamp on the connection record, compute the delta on checkin. One PR.
- [ ] `connection.create_time`: histogram timing the physical DBAPI connect. One PR.
- [ ] `connection.wait_time` + `connection.pending_requests` + `connection.timeouts`: these sit at the acquire seam and share one instrumentation point, so one PR. I can split `wait_time` off if it grows.

All of these have generated factories in `opentelemetry.semconv._incubating.metrics.db_metrics`, so we don't have to worry about hand-rolled instruments.

Three decisions I'd like settled first:

1. Opt-in gating. `connection.count` had a deprecated predecessor (`connections.usage`) to preserve, so it emits the stable name under the `database` / `database/dup` opt-in. The other seven never shipped, so there's no back-compat reason to gate them. I'd still gate them the same way so the metric set stays consistent. Agree, or should the new ones always emit?

2. `idle.min`. SQLAlchemy has no minimum-idle setting like HikariCP's, so there's nothing to read. I'd omit it with a docs note rather than report a constant 0. OK with you?

3. `max` semantics. I'm reading `max` = `pool_size + max_overflow` and `idle.max` = `pool_size`. Pools without a fixed size (`NullPool`, `StaticPool`) don't expose `size()`/`overflow()`, so I'd guard those gauges and skip them for non-QueuePool pools. Does that match what you'd expect?

For tracking, would you rather have separate issues per slice or a task list checked off here in #1158? Either works for me. I'd take them one at a time and pick up the next once #5063 lands, so it'd help if you assign each as I start it rather than all at once.

**Comment by tammy-baylis-swi** [comment]
Thanks @ahmed5145 ! I'm going to confirm some of this with others in the Python SIG at tomorrow's meeting ([feel free to join](https://calendar.google.com/calendar/u/0/event?eid=Mjl0cnRkZjVjYTRxZnBnamMwZHRqaHVjYThfMjAyNjA5MTdUMTYwMDAwWiBjXzJiZjczZTNiNmI1MzBkYTRiYWJkNDQ0ZTcyYjc2YTZhZDg5M2E1YzNmNDNjZjQwNDY3YWJjN2E5YTg5N2Y5NzdAZw) if you're available). I'll add some initial thoughts for now and will review https://github.com/open-telemetry/opentelemetry-python-contrib/pull/5063 soon.

> Remaining slices (each independently reviewable)

These sound good overall. I might recommend an UpDownCounter for max/idle.max instead of a Gauge. We (probably Maintainers) also have to upgrade to a newer semconv version first after [this was recently merged](https://github.com/open-telemetry/semantic-conventions/pull/4073) since semconv calls it [connection.limit](https://github.com/open-telemetry/semantic-conventions/blob/main/model/db/metrics.yaml#L84-L96) but Python still calls it [connection.max](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-semantic-conventions/src/opentelemetry/semconv/_incubating/metrics/db_metrics.py#L77-L91).

> Three decisions I'd like settled first

> The other seven never shipped, so there's no back-compat reason to gate them.

I'm leaning towards this, and not emitting the new ones all the time. But will confirm.

> `idle.min` ... omit it with a docs

Will also confirm.

> Pools without a fixed size (NullPool, StaticPool) don't expose size()/overflow(), so I'd guard those gauges and skip them

This makes sense to me.

**Comment by ahmed5145** [comment]
Thanks @tammy-baylis-swi ! Appreciate the invite for SIG meeting; I have class at that time so I may only be able to lurk, but I'll catch up on whatever got decided during it if there would be notes/recording.

I like your recommendation on the UpDownCounter. It makes more sense; I will use that. And agreed on holding until the semconv version upgrade. 

In the meantime, the name-stable slices aren't blocked, so I'd plan to start with use_time after #5063 lands.

I'll defer to the SIG on the gating and idle.min calls.

**Comment by tammy-baylis-swi** [comment]
Hi @ahmed5145 sounds good! I got some feedback from the Maintainers:

(1) Keep going with the focus on SQLAlchemy first! As you noted there are some metrics we won't be able to generalize in DB-API utility anyway.

(2) Let's simplify all approaches by _not_ using `OTEL_SEMCONV_STABILITY_OPT_IN`, and instead **always emit** the modern metrics -- a pivot from my original idea. The original intent of that env var was for span attribute migration and that doesn't apply to metrics in other instrumentations. No need to backport any "old" metrics. If any other "old" metrics are already being written then they can just be replaced with new, i.e.:

The scope of https://github.com/open-telemetry/opentelemetry-python-contrib/issues/5060 will change and I'll update that ticket next. Then your PR can be cut down to a modernization PR if you like: https://github.com/open-telemetry/opentelemetry-python-contrib/pull/5063

Let me know if any other questions.
